web/main/models.py

```diff
@@ -1093,16 +1112,17 @@ def casebook_fts(
             )
             .values_list("id", "headlines")
         )
-        ids_headlines = {i: h for i, h in ids_headlines}
-        ids_ordinals = None
+
+        ids_headlines = {i: h for i, h in ids_headlines_query}
+        ids_ordinals: Dict[int, List[str]]
         if category == "legal_doc_fulltext":
-            ids_ordinals = (
+            ids_ordinals_nodes = (
                 casebook.contents.filter(resource_type="LegalDocument")
-                .filter(resource_id__in=[r.result_id for r in results])
+                .filter(resource_id__in=[r.result_id for r in results_chunk])
                 .values_list("resource_id", "ordinals")
             )
-            ids_ordinals = {i: [str(n) for n in h] for i, h in ids_ordinals}
-        for r in results:
+            ids_ordinals = {i: [str(n) for n in h] for i, h in ids_ordinals_nodes}  # type: ignore
```

Comment:
I'm a little confused about what's happening [here](https://github.com/harvard-lil/h2o/pull/1639/files#diff-0088ea3fd6dd38dc8969fda2a1559c6de8bab04df5fe137f09208738e9490e97L1096-L1124): the new logic doesn't seem the same to me as the original. Am I reading this wrong, or is this in fact different?

`ids_ordinals` is no longer initialized with `None`; it is now defined inside the `if category == "legal_doc_fulltext":` block. Does this code still work without `ids_ordinals = None`, when ` category != "legal_doc_fulltext"`?

Why is `for r in results` no longer necessary? I see that `results` has been renamed `results_chunk`: does this need to be `for r in results_chunk:` instead of deleted?

Added line 1117 adds a type annotation for `ids_ordinals`, but added line 1124 is marked up with ` # type: ignore`. Can you help me understand why that's the case?

Reply:
> `ids_ordinals` is no longer initialized with `None`; it is now defined inside the if `category == "legal_doc_fulltext":` block. Does this code still work without `ids_ordinals = None`, when `category != "legal_doc_fulltext"`?

Within the for loop, since `id_ordinals` is never touched outside the if block, this code should remain the same.

If `ids_ordinals` was referenced, the following error would be raised: `UnboundLocalError: local variable 'ids_ordinals' referenced before assignment` (rather than a `NameError: name 'ids_ordinals' is not defined`

> I see that `results` has been renamed `results_chunk`:

N.B. `results` is initialized as type `QuerySet` due to `results = base_query.filter(category=category)`. Thus, attempting to assign `result` to `Paginator(results, page_size).get_page(page)` a few lines down will result in a type error since `.get_page` return a `Page`, rather than a `QuerySet` object. So the `results` remains, but only as a `QuerySet` object.

>  does this need to be for `r in results_chunk:` instead of deleted?

Er, sorry, not sure what you mean, `for r in results` was [replaced ](https://github.com/harvard-lil/h2o/pull/1639/files#diff-0088ea3fd6dd38dc8969fda2a1559c6de8bab04df5fe137f09208738e9490e97L1105-R1125)with `for r in results_chunk`. 

![image](https://user-images.githubusercontent.com/8096144/180343481-15ecdee6-6449-4ccd-835d-ea40a2e29659.png)

> Added line 1117 adds a type annotation for `ids_ordinals`, but added line 1124 is marked up with `# type: ignore`. Can you help me understand why that's the case?

Since `ContentNode.resource_id` is a field with attribute `null=True`, so even though all `ContentNode`'s of `resource_type` `LegalDocument` all have `resource_id`, `mypy` isn't aware of that, and will throw an error. However, the comment is an opaque fix and there is probably a better way to indicate what's going on here.

Reply:
Okay, now I see how this works: `ids_ordinals` can be defined exclusively `if category == "legal_doc_fulltext"` because it is only used `if category == "legal_doc_fulltext"`....Something about the way the diff was displaying in Github was throwing me, don't know why. And similarly, sorry about "does this need to be for r in results_chunk: instead of deleted?": that too was me misreading the diff.

Reply:
Re: `ContentNode.resource_id` and the error...
```
main/models.py:1122: error: Key expression in dictionary comprehension has incompatible type "Optional[int]"; expected type "int"
```
... got it. 

If you want the annotation to be [`ids_ordinals: dict[int, list[str]]`](https://github.com/harvard-lil/h2o/pull/1639/files#diff-0088ea3fd6dd38dc8969fda2a1559c6de8bab04df5fe137f09208738e9490e97R1117) instead of `ids_ordinals: dict[Optional[int], list[str]]`, you can explicitly check that `i` is truthy in the dict comprehension:
```
ids_ordinals = {i: [str(n) for n in h] for i, h in ids_ordinals_nodes if i}
```

That seems less desirable to me than using the signature `ids_ordinals: dict[Optional[int], list[str]]`, or not annotating `ids_ordinals` at all.

I think it would also be better to annotate `ids_ordinals` where is defined, rather than above the `if`.

I'm going to make bold and push a commit that switches to `Optional[int]` and puts the annotation inside the `if`; please feel free to lobby for something else.